File: zig/src/core/porter.zig
const std = @import("std");
const ascii = @import("ascii.zig");
const types = @import("types.zig");

fn isConsonant(word: []const u8, i: usize) bool {
    const ch = word[i];
    return switch (ch) {
        'a', 'e', 'i', 'o', 'u' => false,
        'y' => if (i == 0) true else !isConsonant(word, i - 1),
        else => true,
    };
}

fn measure(word: []const u8, len: usize) usize {
    var m: usize = 0;
    var i: usize = 0;

    while (i < len) {
        while (i < len and isConsonant(word, i)) : (i += 1) {}
        if (i >= len) break;
        while (i < len and !isConsonant(word, i)) : (i += 1) {}
        if (i >= len) break;
        m += 1;
    }

    return m;
}

fn containsVowel(word: []const u8, len: usize) bool {
    for (0..len) |i| {
        if (!isConsonant(word, i)) return true;
    }
    return false;
}

fn endsWith(word: []const u8, len: usize, suffix: []const u8) bool {
    if (len < suffix.len) return false;
    return std.mem.eql(u8, word[len - suffix.len .. len], suffix);
}

fn isDoubleConsonant(word: []const u8, len: usize) bool {
    if (len < 2) return false;
    if (word[len - 1] != word[len - 2]) return false;
    return isConsonant(word, len - 1);
}

fn isCvc(word: []const u8, len: usize) bool {
    if (len < 3) return false;

    const c1 = isConsonant(word, len - 1);
    const v = !isConsonant(word, len - 2);
    const c0 = isConsonant(word, len - 3);
    if (!(c0 and v and c1)) return false;

    const ch = word[len - 1];
    return !(ch == 'w' or ch == 'x' or ch == 'y');
}

fn appendE(word: []u8, len_ptr: *usize) void {
    word[len_ptr.*] = 'e';
    len_ptr.* += 1;
}

fn replaceSuffix(word: []u8, len_ptr: *usize, suffix: []const u8, repl: []const u8) void {
    const stem_len = len_ptr.* - suffix.len;
    @memcpy(word[stem_len .. stem_len + repl.len], repl);
    len_ptr.* = stem_len + repl.len;
}

fn step1a(word: []u8, len_ptr: *usize) void {
    if (endsWith(word, len_ptr.*, "sses")) {
        len_ptr.* -= 2;
    } else if (endsWith(word, len_ptr.*, "ies")) {
        len_ptr.* -= 2;
    } else if (endsWith(word, len_ptr.*, "ss")) {
        return;
    } else if (endsWith(word, len_ptr.*, "s")) {
        len_ptr.* -= 1;
    }
}

fn step1bPost(word: []u8, len_ptr: *usize) void {
    if (endsWith(word, len_ptr.*, "at") or endsWith(word, len_ptr.*, "bl") or endsWith(word, len_ptr.*, "iz")) {
        appendE(word, len_ptr);
    } else if (isDoubleConsonant(word, len_ptr.*)) {
        const last = word[len_ptr.* - 1];
        if (!(last == 'l' or last == 's' or last == 'z')) {
            len_ptr.* -= 1;
        }
    } else if (measure(word, len_ptr.*) == 1 and isCvc(word, len_ptr.*)) {
        appendE(word, len_ptr);
    }
}

fn step1b(word: []u8, len_ptr: *usize) void {
    if (endsWith(word, len_ptr.*, "eed")) {
        const stem = len_ptr.* - 3;
        if (measure(word, stem) > 0) len_ptr.* -= 1;
        return;
    }

    if (endsWith(word, len_ptr.*, "ed")) {
        const stem = len_ptr.* - 2;
        if (containsVowel(word, stem)) {
            len_ptr.* = stem;
            step1bPost(word, len_ptr);
        }
        return;
    }

    if (endsWith(word, len_ptr.*, "ing")) {
        const stem = len_ptr.* - 3;
        if (containsVowel(word, stem)) {
            len_ptr.* = stem;
            step1bPost(word, len_ptr);
        }
    }
}

fn step1c(word: []u8, len_ptr: *usize) void {
    if (len_ptr.* == 0) return;
    if (endsWith(word, len_ptr.*, "y") and containsVowel(word, len_ptr.* - 1)) {
        word[len_ptr.* - 1] = 'i';
    }
}

const Rule = struct {
    suffix: []const u8,
    replacement: []const u8,
};

fn applyRulesWithMinMeasure(word: []u8, len_ptr: *usize, rules: []const Rule, min_measure: usize) void {
    for (rules) |rule| {
        if (!endsWith(word, len_ptr.*, rule.suffix)) continue;
        const stem = len_ptr.* - rule.suffix.len;
        if (measure(word, stem) > min_measure) {
            replaceSuffix(word, len_ptr, rule.suffix, rule.replacement);
        }
        return;
    }
}

fn step2(word: []u8, len_ptr: *usize) void {
    const rules = [_]Rule{
        .{ .suffix = "ational", .replacement = "ate" },
        .{ .suffix = "tional", .replacement = "tion" },
        .{ .suffix = "enci", .replacement = "ence" },
        .{ .suffix = "anci", .replacement = "ance" },
        .{ .suffix = "izer", .replacement = "ize" },
        .{ .suffix = "abli", .replacement = "able" },
        .{ .suffix = "alli", .replacement = "al" },
        .{ .suffix = "entli", .replacement = "ent" },
        .{ .suffix = "eli", .replacement = "e" },
        .{ .suffix = "ousli", .replacement = "ous" },
        .{ .suffix = "ization", .replacement = "ize" },
        .{ .suffix = "ation", .replacement = "ate" },
        .{ .suffix = "ator", .replacement = "ate" },
        .{ .suffix = "alism", .replacement = "al" },
        .{ .suffix = "iveness", .replacement = "ive" },
        .{ .suffix = "fulness", .replacement = "ful" },
        .{ .suffix = "ousness", .replacement = "ous" },
        .{ .suffix = "aliti", .replacement = "al" },
        .{ .suffix = "iviti", .replacement = "ive" },
        .{ .suffix = "biliti", .replacement = "ble" },
        .{ .suffix = "logi", .replacement = "log" },
    };
    applyRulesWithMinMeasure(word, len_ptr, &rules, 0);
}

fn step3(word: []u8, len_ptr: *usize) void {
    const rules = [_]Rule{
        .{ .suffix = "icate", .replacement = "ic" },
        .{ .suffix = "ative", .replacement = "" },
        .{ .suffix = "alize", .replacement = "al" },
        .{ .suffix = "iciti", .replacement = "ic" },
        .{ .suffix = "ical", .replacement = "ic" },
        .{ .suffix = "ful", .replacement = "" },
        .{ .suffix = "ness", .replacement = "" },
    };
    applyRulesWithMinMeasure(word, len_ptr, &rules, 0);
}

fn step4(word: []u8, len_ptr: *usize) void {
    const rules = [_][]const u8{
        "ement", "ance", "ence", "able", "ible", "ment", "ant", "ent", "ism", "ate", "iti", "ous", "ive", "ize", "al", "er", "ic", "ou",
    };

    for (rules) |suffix| {
        if (!endsWith(word, len_ptr.*, suffix)) continue;
        const stem = len_ptr.* - suffix.len;
        if (measure(word, stem) > 1) {
            len_ptr.* = stem;
        }
        return;
    }

    if (endsWith(word, len_ptr.*, "ion")) {
        const stem = len_ptr.* - 3;
        if (measure(word, stem) > 1 and stem > 0) {
            const ch = word[stem - 1];
            if (ch == 's' or ch == 't') {
                len_ptr.* = stem;
            }
        }
    }
}

fn step5a(word: []u8, len_ptr: *usize) void {
    if (!endsWith(word, len_ptr.*, "e")) return;
    const stem = len_ptr.* - 1;
    const m = measure(word, stem);
    if (m > 1 or (m == 1 and !isCvc(word, stem))) {
        len_ptr.* = stem;
    }
}

fn step5b(word: []u8, len_ptr: *usize) void {
    if (measure(word, len_ptr.*) > 1 and isDoubleConsonant(word, len_ptr.*) and endsWith(word, len_ptr.*, "l")) {
        len_ptr.* -= 1;
    }
}

pub fn stemPorterAscii(input: []const u8, output: []u8) types.CountError!usize {
    if (output.len < input.len) return error.InsufficientCapacity;
    if (input.len == 0) return 0;

    var len = input.len;
    for (input, 0..) |ch, i| {
        output[i] = ascii.asciiLower(ch);
    }

    if (len <= 2) return len;

    step1a(output, &len);
    step1b(output, &len);
    step1c(output, &len);
    step2(output, &len);
    step3(output, &len);
    step4(output, &len);
    step5a(output, &len);
    step5b(output, &len);

    return len;
}

test "porter sample vectors" {
    const samples = [_][2][]const u8{
        .{ "caresses", "caress" },
        .{ "ponies", "poni" },
        .{ "ties", "ti" },
        .{ "cats", "cat" },
        .{ "feed", "feed" },
        .{ "agreed", "agre" },
        .{ "plastered", "plaster" },
        .{ "motoring", "motor" },
        .{ "sing", "sing" },
        .{ "conflated", "conflat" },
        .{ "hopping", "hop" },
        .{ "filing", "file" },
        .{ "happy", "happi" },
        .{ "sky", "sky" },
        .{ "relational", "relat" },
        .{ "triplicate", "triplic" },
        .{ "probate", "probat" },
        .{ "rate", "rate" },
        .{ "controll", "control" },
        .{ "roll", "roll" },
    };

    var buf = [_]u8{0} ** 64;
    for (samples) |pair| {
        const word = pair[0];
        const expected = pair[1];
        const got_len = try stemPorterAscii(word, &buf);
        try std.testing.expectEqualStrings(expected, buf[0..got_len]);
    }
}

File: rust/src/porter.rs
use crate::ascii;
use crate::CoreError;
use crate::CoreResult;

fn is_consonant(word: &[u8], i: usize) -> bool {
    match word[i] {
        b'a' | b'e' | b'i' | b'o' | b'u' => false,
        b'y' => {
            if i == 0 {
                true
            } else {
                !is_consonant(word, i - 1)
            }
        }
        _ => true,
    }
}

fn measure(word: &[u8], len: usize) -> usize {
    let mut m: usize = 0;
    let mut i: usize = 0;

    while i < len {
        while i < len && is_consonant(word, i) {
            i += 1;
        }
        if i >= len {
            break;
        }
        while i < len && !is_consonant(word, i) {
            i += 1;
        }
        if i >= len {
            break;
        }
        m += 1;
    }

    m
}

fn contains_vowel(word: &[u8], len: usize) -> bool {
    for i in 0..len {
        if !is_consonant(word, i) {
            return true;
        }
    }
    false
}

fn ends_with(word: &[u8], len: usize, suffix: &[u8]) -> bool {
    if len < suffix.len() {
        return false;
    }
    &word[len - suffix.len()..len] == suffix
}

fn is_double_consonant(word: &[u8], len: usize) -> bool {
    if len < 2 {
        return false;
    }
    if word[len - 1] != word[len - 2] {
        return false;
    }
    is_consonant(word, len - 1)
}

fn is_cvc(word: &[u8], len: usize) -> bool {
    if len < 3 {
        return false;
    }

    let c1 = is_consonant(word, len - 1);
    let v = !is_consonant(word, len - 2);
    let c0 = is_consonant(word, len - 3);
    if !(c0 && v && c1) {
        return false;
    }

    let ch = word[len - 1];
    !(ch == b'w' || ch == b'x' || ch == b'y')
}

fn append_e(word: &mut [u8], len: &mut usize) {
    word[*len] = b'e';
    *len += 1;
}

fn replace_suffix(word: &mut [u8], len: &mut usize, suffix_len: usize, repl: &[u8]) {
    let stem_len = *len - suffix_len;
    word[stem_len..stem_len + repl.len()].copy_from_slice(repl);
    *len = stem_len + repl.len();
}

fn step1a(word: &mut [u8], len: &mut usize) {
    if ends_with(word, *len, b"sses") {
        *len -= 2;
    } else if ends_with(word, *len, b"ies") {
        *len -= 2;
    } else if ends_with(word, *len, b"ss") {
        return;
    } else if ends_with(word, *len, b"s") {
        *len -= 1;
    }
}

fn step1b_post(word: &mut [u8], len: &mut usize) {
    if ends_with(word, *len, b"at") || ends_with(word, *len, b"bl") || ends_with(word, *len, b"iz") {
        append_e(word, len);
    } else if is_double_consonant(word, *len) {
        let last = word[*len - 1];
        if !(last == b'l' || last == b's' || last == b'z') {
            *len -= 1;
        }
    } else if measure(word, *len) == 1 && is_cvc(word, *len) {
        append_e(word, len);
    }
}

fn step1b(word: &mut [u8], len: &mut usize) {
    if ends_with(word, *len, b"eed") {
        let stem = *len - 3;
        if measure(word, stem) > 0 {
            *len -= 1;
        }
        return;
    }

    if ends_with(word, *len, b"ed") {
        let stem = *len - 2;
        if contains_vowel(word, stem) {
            *len = stem;
            step1b_post(word, len);
        }
        return;
    }

    if ends_with(word, *len, b"ing") {
        let stem = *len - 3;
        if contains_vowel(word, stem) {
            *len = stem;
            step1b_post(word, len);
        }
    }
}

fn step1c(word: &mut [u8], len: &mut usize) {
    if *len == 0 {
        return;
    }
    if ends_with(word, *len, b"y") && contains_vowel(word, *len - 1) {
        word[*len - 1] = b'i';
    }
}

struct Rule {
    suffix: &'static [u8],
    replacement: &'static [u8],
}

fn apply_rules_with_min_measure(word: &mut [u8], len: &mut usize, rules: &[Rule], min_measure: usize) {
    for rule in rules {
        if !ends_with(word, *len, rule.suffix) {
            continue;
        }
        let stem = *len - rule.suffix.len();
        if measure(word, stem) > min_measure {
            replace_suffix(word, len, rule.suffix.len(), rule.replacement);
        }
        return;
    }
}

fn step2(word: &mut [u8], len: &mut usize) {
    const RULES: [Rule; 21] = [
        Rule { suffix: b"ational", replacement: b"ate" },
        Rule { suffix: b"tional", replacement: b"tion" },
        Rule { suffix: b"enci", replacement: b"ence" },
        Rule { suffix: b"anci", replacement: b"ance" },
        Rule { suffix: b"izer", replacement: b"ize" },
        Rule { suffix: b"abli", replacement: b"able" },
        Rule { suffix: b"alli", replacement: b"al" },
        Rule { suffix: b"entli", replacement: b"ent" },
        Rule { suffix: b"eli", replacement: b"e" },
        Rule { suffix: b"ousli", replacement: b"ous" },
        Rule { suffix: b"ization", replacement: b"ize" },
        Rule { suffix: b"ation", replacement: b"ate" },
        Rule { suffix: b"ator", replacement: b"ate" },
        Rule { suffix: b"alism", replacement: b"al" },
        Rule { suffix: b"iveness", replacement: b"ive" },
        Rule { suffix: b"fulness", replacement: b"ful" },
        Rule { suffix: b"ousness", replacement: b"ous" },
        Rule { suffix: b"aliti", replacement: b"al" },
        Rule { suffix: b"iviti", replacement: b"ive" },
        Rule { suffix: b"biliti", replacement: b"ble" },
        Rule { suffix: b"logi", replacement: b"log" },
    ];
    apply_rules_with_min_measure(word, len, &RULES, 0);
}

fn step3(word: &mut [u8], len: &mut usize) {
    const RULES: [Rule; 7] = [
        Rule { suffix: b"icate", replacement: b"ic" },
        Rule { suffix: b"ative", replacement: b"" },
        Rule { suffix: b"alize", replacement: b"al" },
        Rule { suffix: b"iciti", replacement: b"ic" },
        Rule { suffix: b"ical", replacement: b"ic" },
        Rule { suffix: b"ful", replacement: b"" },
        Rule { suffix: b"ness", replacement: b"" },
    ];
    apply_rules_with_min_measure(word, len, &RULES, 0);
}

fn step4(word: &mut [u8], len: &mut usize) {
    const RULES: [&[u8]; 18] = [
        b"ement", b"ance", b"ence", b"able", b"ible", b"ment", b"ant", b"ent", b"ism", b"ate",
        b"iti", b"ous", b"ive", b"ize", b"al", b"er", b"ic", b"ou",
    ];

    for suffix in RULES.iter() {
        if !ends_with(word, *len, suffix) {
            continue;
        }
        let stem = *len - suffix.len();
        if measure(word, stem) > 1 {
            *len = stem;
        }
        return;
    }

    if ends_with(word, *len, b"ion") {
        let stem = *len - 3;
        if measure(word, stem) > 1 && stem > 0 {
            let ch = word[stem - 1];
            if ch == b's' || ch == b't' {
                *len = stem;
            }
        }
    }
}

fn step5a(word: &mut [u8], len: &mut usize) {
    if !ends_with(word, *len, b"e") {
        return;
    }
    let stem = *len - 1;
    let m = measure(word, stem);
    if m > 1 || (m == 1 && !is_cvc(word, stem)) {
        *len = stem;
    }
}

fn step5b(word: &mut [u8], len: &mut usize) {
    if measure(word, *len) > 1 && is_double_consonant(word, *len) && ends_with(word, *len, b"l") {
        *len -= 1;
    }
}

pub fn stem_porter_ascii(input: &[u8], output: &mut [u8]) -> CoreResult<usize> {
    if output.len() < input.len() {
        return Err(CoreError::InsufficientCapacity);
    }
    if input.is_empty() {
        return Ok(0);
    }

    let mut len = input.len();
    for (i, &ch) in input.iter().enumerate() {
        output[i] = ascii::ascii_lower(ch);
    }

    if len <= 2 {
        return Ok(len);
    }

    step1a(output, &mut len);
    step1b(output, &mut len);
    step1c(output, &mut len);
    step2(output, &mut len);
    step3(output, &mut len);
    step4(output, &mut len);
    step5a(output, &mut len);
    step5b(output, &mut len);

    Ok(len)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn porter_sample_vectors() {
        const SAMPLES: [(&[u8], &[u8]); 20] = [
            (b"caresses", b"caress"),
            (b"ponies", b"poni"),
            (b"ties", b"ti"),
            (b"cats", b"cat"),
            (b"feed", b"feed"),
            (b"agreed", b"agre"),
            (b"plastered", b"plaster"),
            (b"motoring", b"motor"),
            (b"sing", b"sing"),
            (b"conflated", b"conflat"),
            (b"hopping", b"hop"),
            (b"filing", b"file"),
            (b"happy", b"happi"),
            (b"sky", b"sky"),
            (b"relational", b"relat"),
            (b"triplicate", b"triplic"),
            (b"probate", b"probat"),
            (b"rate", b"rate"),
            (b"controll", b"control"),
            (b"roll", b"roll"),
        ];

        let mut buf = [0u8; 64];
        for (word, expected) in SAMPLES.iter() {
            let got_len = stem_porter_ascii(word, &mut buf).unwrap();
            assert_eq!(&buf[0..got_len], *expected);
        }
    }
}
